File: 0076-minimum-window-substring/0076-minimum-window-substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if len(s) < len(t): return ""        
        left = 0
        right = 0
        need = Counter(t)
        missing = len(t)
        minLeft, minRight = 0, 0
        
        while right < len(s):
            missing -= (need[s[right]] > 0)
            need[s[right]] -= 1
            right += 1
            if not missing:
                while left < right and need[s[left]] < 0:
                    need[s[left]] += 1
                    left += 1
                if not minRight or minRight - minLeft > right - left:
                    minLeft, minRight = left, right
            
        return s[minLeft:minRight]
            

File: 0076-minimum-window-substring/0076-minimum-window-substring.cpp
class Solution {
public:
    string minWindow(string& s, string& t) {
        int left = 0, right = 0, minLeft = 0, minRight = 100001, count = t.size();
        int counter[64] = {};
        
        for (char c : t) ++counter[c - 'A'];
        while (right < s.size()) {
            count -= (counter[s[right++] - 'A']-- > 0);
            if (count == 0) {
                while (counter[s[left] - 'A'] < 0) ++counter[s[left++] - 'A'];
                if (minRight - minLeft > right - left) minLeft = left, minRight = right;
            }
        }
        return minRight == 100001 ? "" : s.substr(minLeft, minRight - minLeft);
    }
};